### ISSUE TITLE
Regression: Delayed exceptions for Generators decorated with `not_implemented_for`

### ISSUE THREAD

**Issue by rossbar** [networkx/networkx#5520]
This was originally identified in #5486 - splitting out into a separate issue so we can focus on this specifically.

There has been a behavior change in `not_implemented_for` after NX 2.5. Previously, the `not_implemented_for` would raise exceptions "eagerly" - i.e. behave as if the input checking occurred before any of the body of the decorated function was run. After 2.5, `not_implemented_for` switched to raising "lazily", at least for Generators. In other words, the exception is not raised when the generator is *created*, only when it is iterated over.

## Minimal reproducing example

```python
>>> import networkx as nx
>>> from networkx.utils import not_implemented_for
>>> @not_implemented_for("multigraph")
... def my_gen(G):
...     yield 1
>>> G = nx.MultiGraph()
```

#### In NX 2.5:

```python
>>> mg = my_gen(G)  # Error at generator creation time
Traceback (most recent call last)
   ...
NetworkXNotImplemented: not implemented for multigraph type
```

#### NX > 2.5

```python
>>> mg = my_gen(G)  # Generator created without exception
>>> mg
<generator object my_gen at 0x7f3425964580>
>>> list(mg)  # Exception only raised when Generator values accessed
Traceback (most recent call last)
   ...
NetworkXNotImplemented: not implemented for multigraph type
```

I think the change is related to switching from the `decorator` to `argmap` in 2.6. It could be that this behavior is actually intentional, in which case this issue can be ignored! However I think there are some cases (see e.g. the use-case in #5486) where it is useful to raise the `NetworkXNotImplemented` exception at generator creation-time rather than access-time.

**Comment by boothby** [comment]
This is not intentional.  I'll look for a fix this week.

**Comment by boothby** [comment]
Well, it isn't entirely intentional, anyway.  The problem is here:

```        if inspect.isgeneratorfunction(f):

            def func(*args, __wrapper=None, **kwargs):
                yield from argmap._lazy_compile(__wrapper)(*args, **kwargs)

        else:

            def func(*args, __wrapper=None, **kwargs):
                return argmap._lazy_compile(__wrapper)(*args, **kwargs)
```

That is, if we're decorating a generator, we make a generator.  Therefore, the prequel to the eventual function call is not evaluated until the generator is pumped for the first time.  My first inclination was to add a parameter to `argmap` so that `not_implemented_for` could clobber the generator: 

```        if inspect.isgeneratorfunction(f) or self._clobber_generator:

            def func(*args, __wrapper=None, **kwargs):
                yield from argmap._lazy_compile(__wrapper)(*args, **kwargs)

        else:

            def func(*args, __wrapper=None, **kwargs):
                return argmap._lazy_compile(__wrapper)(*args, **kwargs)
```

However, that would make the behavior position-dependent because of the flattening behavior -- in the code below, `foo` would delay its exception and `bar` would raise it immediately.

```python
@open_file('bla')
@not_implemented_for(digraph)
def foo(...):
   yield ...

@not_implemented_for(digraph)
@open_file('bla')
def bar(...):
    yield ...
```

I'm tempted to remove the generator-dependent behavior altogether.  I think I got the idea from `decorator`, but I don't know why it's important.  And upon revisiting the code, I suspect that `yield from` is actually slower than `return`.  I'll time that and report back.

**Comment by boothby** [comment]
```
Python 3.8.10 (default, Mar 15 2022, 12:22:08) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.13.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: def foo(): 
   ...:     return baz() 
   ...: def bar(): 
   ...:     yield from baz() 
   ...: def baz(): 
   ...:     yield 0; yield 1; yield 2; yield 3; yield 4; yield 5; yield 6; yield 7; yield 8; yield 9 
   ...:                                                                                                                                     

In [2]: %timeit [*foo()]                                                                                                                    
317 ns ± 0.967 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [3]: %timeit [*bar()]                                                                                                                    
507 ns ± 9.25 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

My *extremely* scientific benchmarking confirms my suspicions.  We can get both (miniscule) runtime and (trifling) compile-time performance boost by removing the generator-dependent behavior.  The only reason I can imagine wanting to maintain this code is if `argmap` was doing something super expensive and/or we wanted perfect laziness.  I can't think of anything that we'd need that for in the library... @rossbar @jarrodmillman are you aware of any?

**Comment by boothby** [comment]
Alternatively we could stop flattening across `not_implemented_for`, moving the `self._clobber_generator` check out of `__call__` and into `self.assemble`.  There is some position-dependence, which seems okay; the only issue I see is if somebody expects `open_file` to happen lazily but opens all the files in a large array, they could exceed the file descriptor limit.  To which, I think the answer is to do it lazily and don't make an array?

**Comment by boothby** [comment]
> My vote would be for the simplest (IYO) solution!

Yeah, that's my preference too; I'm mostly documenting my thought process (with invitation to participate) before I bring the hatchet down.

**Comment by boothby** [comment]
Aha, there is some necessary complexity here.  The following unit test fails:

```python
    def test_contextmanager_iterator(self):
        container = []

        def contextmanager(x):
            nonlocal container
            return x, lambda: container.append(x)

        @argmap(contextmanager, 0, 1, 2, try_finally=True)
        def foo(x, y, z):
            yield from (x, y, z)

        q = foo("a", "b", "c")
        assert next(q) == "a"
        assert container == [] # AssertionError: assert ['c', 'b', 'a'] == []
        assert next(q) == "b"
        assert container == []
        assert next(q) == "c"
        assert container == []
        with pytest.raises(StopIteration):
            next(q)

        # context exits are called in reverse
        assert container == ["c", "b", "a"]
```

The problem being that `finally` blocks are executed after returning the generator, and not after exhausting the generator.  In terms of the actual decorators we use:

```python
@open_file('logfile')
def algorithm_with_file_logging(G, logfile):
    for v in G:
        print("processing node", v, file=logfile)
        yield v
...
<ipython-input-19-164e90d9b1ad> in <module>
----> 1 for x in algorithm_with_logging(nx.Graph([(1, 2)]), 'foo'): print(x)

<ipython-input-18-788928c53307> in algorithm_with_logging(G, logfile)
      2 def algorithm_with_logging(G, logfile):
      3     for v in G:
----> 4         print("processing node", v, file=logfile)
      5         yield v
      6 

ValueError: I/O operation on closed file
```

This has a fairly fundamental consequence: we can't eagerly raise an exception for a bad file (without some extreme convolutions, I'm sure).  I'll sleep on this.

**Comment by boothby** [comment]
*sigh* it is not as simple as I would have liked.

Code generation for

```python
@not_implemented_for('digraph')
def foo(x):
    yield x
```
was previously equivalent to
```python
def foo(x):
    if x.is_directed():
        raise NetworkXError
    yield x
```
and is now
```python
def foo(x):
    def intern_foo(x):
        yield x
    if x.is_directed():
        raise NetworkXError
    return intern_foo(x)
```

**Comment by dschult** [comment]
I believe this is now fixed in #5537 so I'll close it. Comment here or reopen if there is more to do.
